### PR TITLE
T021 - One-hot encoding

### PR DESCRIPTION
**Pull request by sakshimisra**
Talktorial ID: 021
* Title: The *One-Hot Encoding* concept
* Original authors: Sakshi Misra
* Reviewer(s): Talia B. Kimber, Prof. Dr. Andrea Volkamer, Yonghui Chen
* Date of review: 10/2020 - present

# Content review

* Potential labels or categories : One Hot Encoding, SMILES, sklearn, Keras
* One line summary: Perform one-hot encoding on SMILES structures by implementing own function, by using implementations in sklearn and keras and then compare their performance in terms of time each implementation takes to execute.
* [x] The table of contents reflects the talktorial story-line; order of #, ##, ### headers is correct
* [x] URLs are linked with meaningful words, instead of pasting the URL directly or linking words like `here`.
* [x] I have spell-checked the notebook
* [x] Images have enough resolution to be rendered with quality, without being _too_ heavy.
* [x] All figures have a description
* [x] Markdown cell content is still in-line with code cell output (whenever results are discussed)
* [x] I have checked that cell outputs are not incredibly long (this applies also to `DataFrames`)
* [x] Formatting looks correctly on the Sphinx render (bold, italics, figure placing)

# Code review

* Time it took to execute (approx.):
* [x] Variable and function names follow snake case rules (e.g. `a_variable_name` vs `aVariableName`)
* [x] Spacing follows PEP8 (run Black on the code cells if needed)
* [x] Code line are under 99 characters each (run `black -l 99`)
* [x] Comments are useful and well placed
* [x] There are no unpythonic idioms like `for i in range(len(list))` (see slides)
* [x] All 3rd party dependencies are listed at the top of the notebook
* [x] I have marked all code cell with output referenced in markdown cells with the label `# NBVAL_CHECK_OUTPUT`
* [x] I have identified potential candidates for a code refactor / useful functions
* [x] All `import ...` lines are at the top (practice part) cell, ordered by standard library / 3rd party packages / our own (`teachopencadd.*`)
* [x] I have update the relative paths to absolute paths.
* List here unfamiliar libraries you find in the imports and their intended use:

### CODE REVIEW

**Review by t-kimber — APPROVED** [review]
I think the notebook has improved a lot and it ready :)

File: teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,2926 @@
+{
```
Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Change the title to <code>The *One-Hot Encoding* concept</code>, because <code>T019</code>   is unsuitable for html webpage.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>This picture can't display in both of my computer and laptop. And for all fugures, I suggest resize them by <code>convert</code> from the terminal:</p><p><code>convert example.png -resize 200x100 example.png</code></p><p> Also suggest add all figures using the Markdown notation: </p><p><code>Short figure title</code></p><p> <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>I suggest format the reference as:</p><p><code>- ChEMBL database: "The ChEMBL bioactivity database: an update." ([&lt;i&gt;Nucleic acids research&lt;i&gt; (2014), &lt;b&gt;42.D1&lt;/b&gt;, D1083-D1090](https://doi.org/10.1093/nar/gkt1031))</code></p><p> instead of pasting the URL directly.</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>I suggest to write <code>*Figure 1*:</code> (in italic), and the same for the following <code>Figure</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Should this packages be the top? I'm not sure.</p><p>I saw a rule that all <code>import ...</code> lines are at the top (practice part) cell, ordered by standard library / 3rd party packages / our own (<code>teachopencadd.*</code>)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>the code shoule be:</p><pre class="ql-syntax" spellcheck="false">df = pd.read_csv(DATA/'CHEMBL25_activities_EGFR.csv',
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;lineterminator='\n', index_col=0)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>I suggest to make the function name more easy to understand, like <code>one_hot_matrix</code> ,<code>one_hot_matrix_plot.</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>I suggest the understandable function name <code>string_split</code> , <code>split_strings</code> with more detial.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/54/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Done editing :)</p><p> </p>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Since every notebook has a talktorial number, so for the unity across the notebooks I have included that talktorial number. Once the final number is decided we can change it later ??</p>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Done editing :) But I would like to ask what difference does it make ?</p>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Done :)</p><p> </p>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Done :)</p>

[the remark this answers is not in the document]
Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Edited, could you please check if all the function name makes sense to you :)</p>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Done:)</p><p> </p>

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Could you please check it again ? </p><p>The image is present in the images folder under '_build/html' folder, so it should be seen in the html page.</p><p>Thank you.</p>

Comment (yonghui-cc) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Oh, It's Ok for <code>T019</code>.  In the lasted version, all the title have talktorial number. Sorry for the mistake.</p>

Comment (yonghui-cc) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>I know what happen :)</p><p>The name of the picture is 'logo.PNG', not 'logo.png'.</p>

Comment (yonghui-cc) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>It's different betwee </p><p><code>'DATA/CHEMBL25_activities_EGFR.csv'</code></p><p>and</p><p><code>DATA/'CHEMBL25_activities_EGFR.csv'</code></p><p> </p>

Comment (yonghui-cc) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: <p>Also, I suggest use <code>"</code>  instead of <code>'</code>  for string in the whole notebook.</p>

File: teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb [hunk 2]
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# T019 · The *One-Hot Encoding* concept\n",
```
Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb]: It's right. Please change the title back. Sorry for my mistake. @sakshimisra

File: teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb (hunk 1 above)
Comment (jaimergp) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: Use `black-nb -l 99 <notebook.ipynb>` to autoformat the talktorials, no need to do it manually!

Comment (sakshimisra) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: Done.

Comment (yonghui-cc) [reply teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:1]: The picture still can't be shown in the notebook. Please correct the name of the picture file 'logo.PNG' to 'logo.png' in the folder. @sakshimisra

File: teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb [hunk 3]
```diff
@@ -65,22 +65,22 @@
...
     "     - [**scikit-learn**](https://scikit-learn.org/stable/):\n",
-    "         - [Scikit-learn: Machine Learning in Python](https://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html), Pedregosa _et al._, JMLR 12, pp. 2825-2830, 2011.\n",
-    "         - Jiangang Hao, et al. \"A Review of Scikit-learn Package in Python Programming Language.\" _Journal of Education and Behavioral Statistics_ Volume: 44 issue: 3 (2019), page(s): 348-361. https://doi.org/10.3102/1076998619832248.\n",
+    "     - [Scikit-learn: Machine Learning in Python](https://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html), Pedregosa *et al.*, JMLR 12, pp. 2825-2830, 2011.\n",
+    "         - Jiangang Hao, et al. \"A Review of Scikit-learn Package in Python Programming Language.\" [_Journal of Education and Behavioral Statistics_ Volume: 44 issue: 3 (2019), page(s): 348-361](https://doi.org/10.3102/1076998619832248.)\n",
     "     - [**keras**](https://keras.io/): Book chapter: \"An Introduction to Deep Learning and Keras\" in *Learn Keras for Deep Neural Networks* (2019), page(s):1-16, https://doi.org/10.1007/978-1-4842-4240-7.\n",
```
Comment (yonghui-cc) [inline teachopencadd/talktorials/019_One_Hot_Encoding/talktorial.ipynb:78]: Maybe miss editing this URL.:)

**Comment by yonghui-cc** [comment]
The notebook looks good from my view now.

**Review by AndreaVolkamer — COMMENTED** [review]
Great nb, @sakshimisra. I reviewed the theory part so far only using reviewnb (though having trouble because they are currently not showing here?!?).

**Comment by AndreaVolkamer** [comment]
@sakshimisra well done (thanks for your help @t-kimber)! 
There are only one or two small things that we'll discuss offline. @jaimergp could you go ahead with merging, as soon as one of us gives you the final go! Thanks everyone.

**Comment by jaimergp** [comment]
~How come there are other notebooks affected?~ Ah, this was opened way before the big merge. I'll need some time to go through the conflicts, but I'll have it done this week.

**Comment by AndreaVolkamer** [comment]
@sakshimisra thanks for the update, well done. 

I reviewed the shortened version of the notebook and checked/adapted mainly these points
- consistent usage of one-hot encoding (vs. one hot)
- SMILES in text vs smiles in code
- text and parameters in doc strings, also including/adapting parameters making sure that no global variables are used
- small typos and textual adaptions
- in replacement: excluded Cn since it is rather carbon (C) + aromatic nitrogen (n)
- excluded compound from cvs file that only contained [Cl-]

Final Todos, please:
- Check the website rendering
- Delete temporary notebook
- While blog posts are very illustrative and easy to understand, please include a more sophisticated reference for OHE if possible
- Maybe we can adapt the discussion slightly to comment more on our field (CADD), how OHE can be used, maybe also potential shortcomings of the SMILES  preprocessing, not sure if the NLP comment is completely clear.

Thanks.

**Comment by jaimergp** [comment]
Thanks everybody, especially @sakshimisra of course! Awesome work ❤️ 

I cleaned the history a bit as per #99 and will close this as a result.
